[FLINK-34137] Update CI to test archunit configuration

echauchot opened the pull request:
Update CI to test skiping archunit tests on non-main Flink versions. Test on submodules with archunit tests and without

R: @MartijnVisser 

FYI: @tzulitai

echauchot commented:
FYI, I was busy on another Flink PR. Resuming this PR. It is almost ready for review, I wanted to do a final test of generating a violation on some flink version and not on another.

zentol commented on .github/workflows/_testing.yml:
```diff
@@ -25,19 +25,25 @@ jobs:
     uses: ./.github/workflows/ci.yml
     with:
-      flink_version: 1.16.1
-      connector_branch: ci_utils
```
Why is this being removed?

echauchot replied:
just for testing as the workflow is called "testing" I thought I could use it to test my setup :smile:

zentol commented on .github/workflows/ci.yml:
```diff
@@ -41,8 +41,8 @@ on:
         default: 50
-      run_dependency_convergence:
-        description: "Whether to run the dependency convergence check"
+      run_sanity_checks:
```
This name is too generic and a term we haven't used anywhere else in our CI setup.

`skip_qa_plugins_and_archunit_tests` would be more descriptive, but you can also immediately tell that it is wildly inaccurate because most QA plugins are still being run..

echauchot replied:
ok. This will be solved when separating the test parameters (see other comments)

zentol commented on .github/workflows/ci.yml:
```diff
@@ -41,8 +41,8 @@ on:
         required: false
         type: number
         default: 50
-      run_dependency_convergence:
```
This breaks compatibility which just creates headaches. It's not too difficult to continue supporting this parameter.

echauchot replied:
Yes I totally agree. As I said in [this comment](https://github.com/apache/flink-connector-shared-utils/pull/23#issuecomment-1775394910), I merged the 2 test parameters because I misunderstood a comment you did on slack. If we put them separated again, no more breaking change.

zentol reviewed (changes requested):
Generally speaking I'm not too fond of a general "turn some things off" switch.
I get the idea and benefits it would bring, but IMO we should be very explicit as to what checks have been disabled.

echauchot commented:
> Generally speaking I'm not too fond of a general "turn some things off" switch. I get the idea and benefits it would bring, but IMO we should be very explicit as to what checks have been disabled. Because of that I'd rather see a dedicated skip-archunit-tests parameter.

This is what I wanted to do at first but I merged the two test parameters as you said in a slack discussion about archunit tests:
> We could actually roll this into the run_dependency_convergence option because it has fundamentally the same cause.

But it seems I misunderstood what you meant. I actually also prefer to keep them separated. So, no problem, I'll make them separated

echauchot commented:
@zentol I fixed the PR: mainly separated dep convergence checks and archunit tests.

I have also simulated an archunit violation with the behavior of the jira ticket: a rule was fixed  in a snapshot leading to a new violation (ConnectorRules#CONNECTOR_CLASSES_ONLY_DEPEND_ON_PUBLIC_API that I fixed and that you reviewed). See [this commit](https://github.com/apache/flink-connector-shared-utils/pull/23/commits/eb28d4ba4f5434e356fd09120c9420f726004266). I think we can leave such a violation in the test project to serve for future tests.

After playing a lot with this simulation locally, I find that the comments in `archunit.properties` could be improved because not precise enough and even a bit misleading. I propose [this comments change](https://github.com/apache/flink-connector-shared-utils/pull/23/commits/a16852321a829fc0c42271570563655c41dc8d14). If you agree with the change, I'll do a PR to flink-architecture-tests module.

snuyanzin commented on .github/workflows/ci.yml:
```diff
@@ -88,6 +93,10 @@ jobs:
 
+      - name: "Disable archunit tests"
+        if: ${{ inputs.skip_archunit_tests }}
+        run: echo "MVN_ARCHUNIT_TESTS=-Dtest='!*ArchitectureTest'" >> $GITHUB_ENV
```
I tried similar settings locally like
`mvn -Dflink.version=1.18.0  -Dtest=\!*ArchitectureTest* clean install` and for dirs without archunit tests it fails like
```
No tests were executed!  (Set -DfailIfNoTests=false to ignore this error.) 
```
then if we add mentioned flag to the command then surefire mixes integration and non integration tests (looks like a bug on surefire side, which does not happen without this flag). The problem is that there is for instance test called `PackagingITCase` assuming that there is already a built jar... however if surefire runs it together with default tests instead of integration then this test fails

would be great if this is somehow covered here...

echauchot replied:
@snuyanzin thanks for your feedback, I'll take a look.

snuyanzin replied:
the idea which seems working at least for my setup with opensearch connector is using `<include>`, `<exclude>` for surefire plugin and different profiles like
```xml
<properties>
    	<exclude.tests>nothing-to-exclude</exclude.tests>
</properties>
...
<plugin>
		<groupId>org.apache.maven.plugins</groupId>
		<artifactId>maven-surefire-plugin</artifactId>
		<configuration>
			...
		</configuration>
		<executions>
			<execution>
                           <!--execute all the unit tests-->
			   <id>default-test</id>			
                            <configuration>
                            
	                            <includes>
		                            <include>${test.unit.pattern}</include>
	                            </includes>
	                            <excludes>
		                            <exclude>${exclude.tests}</exclude>
	                            </excludes>
                                   ...
                            </configuration>
                      ...		
	              </execution>
	              <execution>
		      <!--execute all the integration tests-->
		      <id>integration-tests</id>
		      ...
		      <configuration>
			      <includes>
				      <include>**/*.*</include>
			      </includes>
			      <excludes>
				      <exclude>${test.unit.pattern}</exclude>
				      <exclude>${exclude.tests}</exclude>
			      </excludes>
			      ...
		      </configuration>
	      </execution>
              </executions>
</plugin>
...
	<profiles>
		<profile>
			<id>noArchUnit</id>
			<properties>
				<exclude.tests>**/org/apache/flink/architecture/**/*.java</exclude.tests>
			</properties>
		</profile>
	</profiles>
...
```

echauchot replied:
First of all `-Dtest` is not a surefire parameter but a maven one. 

Then, the above command is not meant to be run from anywhere but from the CI at the root of a connector module that has the flink standard connector structure (the one of test project sample) including the architecture tests.

To be sure, I ran  `mvn install -Dflink.version=1.18.0 -Dtest='!*ArchitectureTest'` from [the cassandra connector root](https://github.com/apache/flink-connector-cassandra/tree/main/flink-connector-cassandra):

1. with all tests in the structure:
=> UTests and ITests are run and archunit tests are skipped as expected (see log [1] below)

2. after removing the archunit tests as you mentioned something about dirs without these tests:
=> UTests and ITests are run as expected (see log [2] below)

==> I cannot reproduce what you observed


[1]
```
[INFO] Scanning for projects...
[WARNING] 
[WARNING] Some problems were encountered while building the effective model for org.apache.flink:flink-connector-cassandra_2.12:jar:4.0-SNAPSHOT
[WARNING] 'artifactId' contains an expression but should be a constant. @ org.apache.flink:flink-connector-cassandra_${scala.binary.version}:4.0-SNAPSHOT, /home/echauchot/projects/flink-connector-cassandra/flink-connector-cassandra/pom.xml, line 32, column 14
[WARNING] 
[WARNING] It is highly recommended to fix these problems because they threaten the stability of your build.
[WARNING] 
[WARNING] For this reason, future Maven versions might no longer support building such malformed projects.
[WARNING] 
[INFO] 
[INFO] ----------< org.apache.flink:flink-connector-cassandra_2.12 >-----------
[INFO] Building Flink : Connectors : Cassandra 4.0-SNAPSHOT
[INFO] --------------------------------[ jar ]---------------------------------
[INFO] 
[INFO] --- maven-checkstyle-plugin:3.1.2:check (validate) @ flink-connector-cassandra_2.12 ---
[WARNING] Old version of checkstyle detected. Consider updating to >= v8.30
[WARNING] For more information see: https://maven.apache.org/plugins/maven-checkstyle-plugin/examples/upgrading-checkstyle.html
[INFO] You have 0 Checkstyle violations.
[INFO] 
[INFO] --- spotless-maven-plugin:2.27.1:check (spotless-check) @ flink-connector-cassandra_2.12 ---
[INFO] 
[INFO] --- maven-enforcer-plugin:3.0.0-M1:enforce (enforce-maven) @ flink-connector-cassandra_2.12 ---
[INFO] 
[INFO] --- maven-enforcer-plugin:3.0.0-M1:enforce (forbid-direct-table-planner-dependencies) @ flink-connector-cassandra_2.12 ---
[INFO] 
[INFO] --- maven-enforcer-plugin:3.0.0-M1:enforce (enforce-maven-version) @ flink-connector-cassandra_2.12 ---
[INFO] 
[INFO] --- maven-remote-resources-plugin:1.5:process (process-resource-bundles) @ flink-connector-cassandra_2.12 ---
[INFO] 
[INFO] --- maven-resources-plugin:3.1.0:resources (default-resources) @ flink-connector-cassandra_2.12 ---
[INFO] Using 'UTF-8' encoding to copy filtered resources.
[INFO] Copying 0 resource
[INFO] Copying 3 resources
[INFO] 
[INFO] --- maven-compiler-plugin:3.8.0:compile (default-compile) @ flink-connector-cassandra_2.12 ---
[INFO] Nothing to compile - all classes are up to date
[INFO] 
[INFO] --- maven-resources-plugin:3.1.0:testResources (default-testResources) @ flink-connector-cassandra_2.12 ---
[INFO] Using 'UTF-8' encoding to copy filtered resources.
[INFO] Copying 3 resources
[INFO] Copying 3 resources
[INFO] 
[INFO] --- maven-compiler-plugin:3.8.0:testCompile (default-testCompile) @ flink-connector-cassandra_2.12 ---
[INFO] Nothing to compile - all classes are up to date
[INFO] 
[INFO] --- maven-surefire-plugin:3.0.0-M5:test (default-test) @ flink-connector-cassandra_2.12 ---
[INFO] 
[INFO] -------------------------------------------------------
[INFO]  T E S T S
[INFO] -------------------------------------------------------
[INFO] Running org.apache.flink.streaming.connectors.cassandra.CassandraTupleWriteAheadSinkTest
[INFO] Running org.apache.flink.streaming.connectors.cassandra.CassandraSinkBaseTest
[INFO] Running org.apache.flink.streaming.connectors.cassandra.CassandraConnectorITCase
[INFO] Running org.apache.flink.connector.cassandra.source.CassandraSourceITCase
[INFO] Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 1.283 s - in org.apache.flink.streaming.connectors.cassandra.CassandraTupleWriteAheadSinkTest
[INFO] Running org.apache.flink.connector.cassandra.source.split.CassandraSplitSerializerTest
[INFO] Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.013 s - in org.apache.flink.connector.cassandra.source.split.CassandraSplitSerializerTest
[INFO] Running org.apache.flink.connector.cassandra.source.enumerator.CassandraEnumeratorStateSerializerTest
[INFO] Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.013 s - in org.apache.flink.connector.cassandra.source.enumerator.CassandraEnumeratorStateSerializerTest
[INFO] Running org.apache.flink.connector.cassandra.source.reader.CassandraQueryTest
[INFO] Tests run: 3, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.02 s - in org.apache.flink.connector.cassandra.source.reader.CassandraQueryTest
[INFO] Tests run: 12, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 1.436 s - in org.apache.flink.streaming.connectors.cassandra.CassandraSinkBaseTest
[ERROR] WARNING: An illegal reflective access operation has occurred
[ERROR] WARNING: Illegal reflective access by org.jboss.netty.util.internal.ByteBufferUtil (file:/tmp/flink-rpc-akkaf91156d2-7ae8-4d71-bd4f-dced7d4a75e1.jar) to method java.nio.DirectByteBuffer.cleaner()
[ERROR] WARNING: Please consider reporting this to the maintainers of org.jboss.netty.util.internal.ByteBufferUtil
[ERROR] WARNING: Use --illegal-access=warn to enable warnings of further illegal reflective access operations
[ERROR] WARNING: All illegal access operations will be denied in a future release
[INFO] Tests run: 7, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 56.932 s - in org.apache.flink.connector.cassandra.source.CassandraSourceITCase
[INFO] Tests run: 18, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 105.701 s - in org.apache.flink.streaming.connectors.cassandra.CassandraConnectorITCase
[INFO] 
[INFO] Results:
[INFO] 
[INFO] Tests run: 43, Failures: 0, Errors: 0, Skipped: 0
[INFO] 
[INFO] 
[INFO] --- maven-jar-plugin:2.4:jar (default-jar) @ flink-connector-cassandra_2.12 ---
[INFO] Building jar: /home/echauchot/projects/flink-connector-cassandra/flink-connector-cassandra/target/flink-connector-cassandra_2.12-4.0-SNAPSHOT.jar
[INFO] 
[INFO] --- maven-shade-plugin:3.2.4:shade (shade-flink) @ flink-connector-cassandra_2.12 ---
[INFO] Including com.datastax.cassandra:cassandra-driver-core:jar:shaded:3.11.2 in the shaded jar.
[INFO] Including com.datastax.cassandra:cassandra-driver-mapping:jar:3.11.2 in the shaded jar.
[INFO] Including com.google.guava:guava:jar:19.0 in the shaded jar.
[INFO] Excluding io.dropwizard.metrics:metrics-core:jar:3.2.2 from the shaded jar.
[INFO] No artifact matching filter org.apache.flink:flink-shaded-force-shading
[INFO] No artifact matching filter io.netty:netty
[WARNING] cassandra-driver-core-3.11.2-shaded.jar, cassandra-driver-mapping-3.11.2.jar, flink-connector-cassandra_2.12-4.0-SNAPSHOT.jar, guava-19.0.jar define 1 overlapping resource: 
[WARNING]   - META-INF/MANIFEST.MF
[WARNING] maven-shade-plugin has detected that some class files are
[WARNING] present in two or more JARs. When this happens, only one
[WARNING] single version of the class is copied to the uber jar.
[WARNING] Usually this is not harmful and you can skip these warnings,
[WARNING] otherwise try to manually exclude artifacts based on
[WARNING] mvn dependency:tree -Ddetail=true and the above output.
[WARNING] See http://maven.apache.org/plugins/maven-shade-plugin/
[INFO] Replacing original artifact with shaded artifact.
[INFO] Replacing /home/echauchot/projects/flink-connector-cassandra/flink-connector-cassandra/target/flink-connector-cassandra_2.12-4.0-SNAPSHOT.jar with /home/echauchot/projects/flink-connector-cassandra/flink-connector-cassandra/target/flink-connector-cassandra_2.12-4.0-SNAPSHOT-shaded.jar
[INFO] Dependency-reduced POM written at: /home/echauchot/projects/flink-connector-cassandra/flink-connector-cassandra/target/dependency-reduced-pom.xml
[INFO] 
[INFO] --- maven-surefire-plugin:3.0.0-M5:test (integration-tests) @ flink-connector-cassandra_2.12 ---
[INFO] 
[INFO] -------------------------------------------------------
[INFO]  T E S T S
[INFO] -------------------------------------------------------
[INFO] Running org.apache.flink.streaming.connectors.cassandra.CassandraSinkBaseTest
[INFO] Running org.apache.flink.streaming.connectors.cassandra.CassandraConnectorITCase
[INFO] Tests run: 12, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 1.075 s - in org.apache.flink.streaming.connectors.cassandra.CassandraSinkBaseTest
[INFO] Running org.apache.flink.streaming.connectors.cassandra.CassandraTupleWriteAheadSinkTest
[INFO] Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.975 s - in org.apache.flink.streaming.connectors.cassandra.CassandraTupleWriteAheadSinkTest
[INFO] Running org.apache.flink.connector.cassandra.source.CassandraSourceITCase
[ERROR] WARNING: An illegal reflective access operation has occurred
[ERROR] WARNING: Illegal reflective access by org.jboss.netty.util.internal.ByteBufferUtil (file:/tmp/flink-rpc-akkaf6320afa-293e-4879-8871-c4fc3058a2cd.jar) to method java.nio.DirectByteBuffer.cleaner()
[ERROR] WARNING: Please consider reporting this to the maintainers of org.jboss.netty.util.internal.ByteBufferUtil
[ERROR] WARNING: Use --illegal-access=warn to enable warnings of further illegal reflective access operations
[ERROR] WARNING: All illegal access operations will be denied in a future release
[INFO] Tests run: 7, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 56.696 s - in org.apache.flink.connector.cassandra.source.CassandraSourceITCase
[INFO] Running org.apache.flink.connector.cassandra.source.split.CassandraSplitSerializerTest
[INFO] Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.07 s - in org.apache.flink.connector.cassandra.source.split.CassandraSplitSerializerTest
[INFO] Running org.apache.flink.connector.cassandra.source.enumerator.CassandraEnumeratorStateSerializerTest
[INFO] Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.081 s - in org.apache.flink.connector.cassandra.source.enumerator.CassandraEnumeratorStateSerializerTest
[INFO] Running org.apache.flink.connector.cassandra.source.reader.CassandraQueryTest
[INFO] Tests run: 3, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.085 s - in org.apache.flink.connector.cassandra.source.reader.CassandraQueryTest
[INFO] Tests run: 18, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 105.637 s - in org.apache.flink.streaming.connectors.cassandra.CassandraConnectorITCase
[INFO] 
[INFO] Results:
[INFO] 
[INFO] Tests run: 43, Failures: 0, Errors: 0, Skipped: 0
[INFO] 
[INFO] 
[INFO] --- japicmp-maven-plugin:0.16.0_m325:cmp (default) @ flink-connector-cassandra_2.12 ---
[INFO] Written file '/home/echauchot/projects/flink-connector-cassandra/flink-connector-cassandra/target/${rootDir}/tools/japicmp-output/flink-connector-cassandra_2.12/japicmp/japicmp.diff'.
[INFO] Written file '/home/echauchot/projects/flink-connector-cassandra/flink-connector-cassandra/target/${rootDir}/tools/japicmp-output/flink-connector-cassandra_2.12/japicmp/japicmp.xml'.
[INFO] Written file '/home/echauchot/projects/flink-connector-cassandra/flink-connector-cassandra/target/${rootDir}/tools/japicmp-output/flink-connector-cassandra_2.12/japicmp/japicmp.html'.
[INFO] 
[INFO] --- maven-install-plugin:2.5.2:install (default-install) @ flink-connector-cassandra_2.12 ---
[INFO] Installing /home/echauchot/projects/flink-connector-cassandra/flink-connector-cassandra/target/flink-connector-cassandra_2.12-4.0-SNAPSHOT.jar to /home/echauchot/.m2/repository/org/apache/flink/flink-connector-cassandra_2.12/4.0-SNAPSHOT/flink-connector-cassandra_2.12-4.0-SNAPSHOT.jar
[INFO] Installing /home/echauchot/projects/flink-connector-cassandra/flink-connector-cassandra/target/dependency-reduced-pom.xml to /home/echauchot/.m2/repository/org/apache/flink/flink-connector-cassandra_2.12/4.0-SNAPSHOT/flink-connector-cassandra_2.12-4.0-SNAPSHOT.pom
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  03:45 min
[INFO] Finished at: 2023-10-27T18:27:09+02:00
[INFO] ------------------------------------------------------------------------

```

[2]

```
[INFO] Scanning for projects...
[WARNING] 
[WARNING] Some problems were encountered while building the effective model for org.apache.flink:flink-connector-cassandra_2.12:jar:4.0-SNAPSHOT
[WARNING] 'artifactId' contains an expression but should be a constant. @ org.apache.flink:flink-connector-cassandra_${scala.binary.version}:4.0-SNAPSHOT, /home/echauchot/projects/flink-connector-cassandra/flink-connector-cassandra/pom.xml, line 32, column 14
[WARNING] 
[WARNING] It is highly recommended to fix these problems because they threaten the stability of your build.
[WARNING] 
[WARNING] For this reason, future Maven versions might no longer support building such malformed projects.
[WARNING] 
[INFO] 
[INFO] ----------< org.apache.flink:flink-connector-cassandra_2.12 >-----------
[INFO] Building Flink : Connectors : Cassandra 4.0-SNAPSHOT
[INFO] --------------------------------[ jar ]---------------------------------
[INFO] 
[INFO] --- maven-checkstyle-plugin:3.1.2:check (validate) @ flink-connector-cassandra_2.12 ---
[WARNING] Old version of checkstyle detected. Consider updating to >= v8.30
[WARNING] For more information see: https://maven.apache.org/plugins/maven-checkstyle-plugin/examples/upgrading-checkstyle.html
[INFO] You have 0 Checkstyle violations.
[INFO] 
[INFO] --- spotless-maven-plugin:2.27.1:check (spotless-check) @ flink-connector-cassandra_2.12 ---
[INFO] 
[INFO] --- maven-enforcer-plugin:3.0.0-M1:enforce (enforce-maven) @ flink-connector-cassandra_2.12 ---
[INFO] 
[INFO] --- maven-enforcer-plugin:3.0.0-M1:enforce (forbid-direct-table-planner-dependencies) @ flink-connector-cassandra_2.12 ---
[INFO] 
[INFO] --- maven-enforcer-plugin:3.0.0-M1:enforce (enforce-maven-version) @ flink-connector-cassandra_2.12 ---
[INFO] 
[INFO] --- maven-remote-resources-plugin:1.5:process (process-resource-bundles) @ flink-connector-cassandra_2.12 ---
[INFO] 
[INFO] --- maven-resources-plugin:3.1.0:resources (default-resources) @ flink-connector-cassandra_2.12 ---
[INFO] Using 'UTF-8' encoding to copy filtered resources.
[INFO] Copying 0 resource
[INFO] Copying 3 resources
[INFO] 
[INFO] --- maven-compiler-plugin:3.8.0:compile (default-compile) @ flink-connector-cassandra_2.12 ---
[INFO] Nothing to compile - all classes are up to date
[INFO] 
[INFO] --- maven-resources-plugin:3.1.0:testResources (default-testResources) @ flink-connector-cassandra_2.12 ---
[INFO] Using 'UTF-8' encoding to copy filtered resources.
[INFO] Copying 3 resources
[INFO] Copying 3 resources
[INFO] 
[INFO] --- maven-compiler-plugin:3.8.0:testCompile (default-testCompile) @ flink-connector-cassandra_2.12 ---
[INFO] Nothing to compile - all classes are up to date
[INFO] 
[INFO] --- maven-surefire-plugin:3.0.0-M5:test (default-test) @ flink-connector-cassandra_2.12 ---
[INFO] 
[INFO] -------------------------------------------------------
[INFO]  T E S T S
[INFO] -------------------------------------------------------
[INFO] Running org.apache.flink.streaming.connectors.cassandra.CassandraTupleWriteAheadSinkTest
[INFO] Running org.apache.flink.streaming.connectors.cassandra.CassandraSinkBaseTest
[INFO] Running org.apache.flink.streaming.connectors.cassandra.CassandraConnectorITCase
[INFO] Running org.apache.flink.connector.cassandra.source.CassandraSourceITCase
[INFO] Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 1.446 s - in org.apache.flink.streaming.connectors.cassandra.CassandraTupleWriteAheadSinkTest
[INFO] Running org.apache.flink.connector.cassandra.source.split.CassandraSplitSerializerTest
[INFO] Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.021 s - in org.apache.flink.connector.cassandra.source.split.CassandraSplitSerializerTest
[INFO] Running org.apache.flink.connector.cassandra.source.enumerator.CassandraEnumeratorStateSerializerTest
[INFO] Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.014 s - in org.apache.flink.connector.cassandra.source.enumerator.CassandraEnumeratorStateSerializerTest
[INFO] Running org.apache.flink.connector.cassandra.source.reader.CassandraQueryTest
[INFO] Tests run: 3, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.018 s - in org.apache.flink.connector.cassandra.source.reader.CassandraQueryTest
[INFO] Tests run: 12, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 1.659 s - in org.apache.flink.streaming.connectors.cassandra.CassandraSinkBaseTest
[ERROR] WARNING: An illegal reflective access operation has occurred
[ERROR] WARNING: Illegal reflective access by org.jboss.netty.util.internal.ByteBufferUtil (file:/tmp/flink-rpc-akkabdb2b72e-22fa-4100-ac6b-aa9126fcdcee.jar) to method java.nio.DirectByteBuffer.cleaner()
[ERROR] WARNING: Please consider reporting this to the maintainers of org.jboss.netty.util.internal.ByteBufferUtil
[ERROR] WARNING: Use --illegal-access=warn to enable warnings of further illegal reflective access operations
[ERROR] WARNING: All illegal access operations will be denied in a future release
[INFO] Tests run: 7, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 57.532 s - in org.apache.flink.connector.cassandra.source.CassandraSourceITCase
[INFO] Tests run: 18, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 106.445 s - in org.apache.flink.streaming.connectors.cassandra.CassandraConnectorITCase
[INFO] 
[INFO] Results:
[INFO] 
[INFO] Tests run: 43, Failures: 0, Errors: 0, Skipped: 0
[INFO] 
[INFO] 
[INFO] --- maven-jar-plugin:2.4:jar (default-jar) @ flink-connector-cassandra_2.12 ---
[INFO] Building jar: /home/echauchot/projects/flink-connector-cassandra/flink-connector-cassandra/target/flink-connector-cassandra_2.12-4.0-SNAPSHOT.jar
[INFO] 
[INFO] --- maven-shade-plugin:3.2.4:shade (shade-flink) @ flink-connector-cassandra_2.12 ---
[INFO] Including com.datastax.cassandra:cassandra-driver-core:jar:shaded:3.11.2 in the shaded jar.
[INFO] Including com.datastax.cassandra:cassandra-driver-mapping:jar:3.11.2 in the shaded jar.
[INFO] Including com.google.guava:guava:jar:19.0 in the shaded jar.
[INFO] Excluding io.dropwizard.metrics:metrics-core:jar:3.2.2 from the shaded jar.
[INFO] No artifact matching filter org.apache.flink:flink-shaded-force-shading
[INFO] No artifact matching filter io.netty:netty
[WARNING] cassandra-driver-core-3.11.2-shaded.jar, cassandra-driver-mapping-3.11.2.jar, flink-connector-cassandra_2.12-4.0-SNAPSHOT.jar, guava-19.0.jar define 1 overlapping resource: 
[WARNING]   - META-INF/MANIFEST.MF
[WARNING] maven-shade-plugin has detected that some class files are
[WARNING] present in two or more JARs. When this happens, only one
[WARNING] single version of the class is copied to the uber jar.
[WARNING] Usually this is not harmful and you can skip these warnings,
[WARNING] otherwise try to manually exclude artifacts based on
[WARNING] mvn dependency:tree -Ddetail=true and the above output.
[WARNING] See http://maven.apache.org/plugins/maven-shade-plugin/
[INFO] Replacing original artifact with shaded artifact.
[INFO] Replacing /home/echauchot/projects/flink-connector-cassandra/flink-connector-cassandra/target/flink-connector-cassandra_2.12-4.0-SNAPSHOT.jar with /home/echauchot/projects/flink-connector-cassandra/flink-connector-cassandra/target/flink-connector-cassandra_2.12-4.0-SNAPSHOT-shaded.jar
[INFO] Dependency-reduced POM written at: /home/echauchot/projects/flink-connector-cassandra/flink-connector-cassandra/target/dependency-reduced-pom.xml
[INFO] 
[INFO] --- maven-surefire-plugin:3.0.0-M5:test (integration-tests) @ flink-connector-cassandra_2.12 ---
[INFO] 
[INFO] -------------------------------------------------------
[INFO]  T E S T S
[INFO] -------------------------------------------------------
[INFO] Running org.apache.flink.streaming.connectors.cassandra.CassandraSinkBaseTest
[INFO] Running org.apache.flink.streaming.connectors.cassandra.CassandraConnectorITCase
[INFO] Tests run: 12, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 1.145 s - in org.apache.flink.streaming.connectors.cassandra.CassandraSinkBaseTest
[INFO] Running org.apache.flink.streaming.connectors.cassandra.CassandraTupleWriteAheadSinkTest
[INFO] Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 1.072 s - in org.apache.flink.streaming.connectors.cassandra.CassandraTupleWriteAheadSinkTest
[INFO] Running org.apache.flink.connector.cassandra.source.CassandraSourceITCase
[ERROR] WARNING: An illegal reflective access operation has occurred
[ERROR] WARNING: Illegal reflective access by org.jboss.netty.util.internal.ByteBufferUtil (file:/tmp/flink-rpc-akka20cb269c-4523-4f18-9763-af2c15df2c6a.jar) to method java.nio.DirectByteBuffer.cleaner()
[ERROR] WARNING: Please consider reporting this to the maintainers of org.jboss.netty.util.internal.ByteBufferUtil
[ERROR] WARNING: Use --illegal-access=warn to enable warnings of further illegal reflective access operations
[ERROR] WARNING: All illegal access operations will be denied in a future release
[INFO] Tests run: 7, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 56.508 s - in org.apache.flink.connector.cassandra.source.CassandraSourceITCase
[INFO] Running org.apache.flink.connector.cassandra.source.split.CassandraSplitSerializerTest
[INFO] Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.068 s - in org.apache.flink.connector.cassandra.source.split.CassandraSplitSerializerTest
[INFO] Running org.apache.flink.connector.cassandra.source.enumerator.CassandraEnumeratorStateSerializerTest
[INFO] Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.077 s - in org.apache.flink.connector.cassandra.source.enumerator.CassandraEnumeratorStateSerializerTest
[INFO] Running org.apache.flink.connector.cassandra.source.reader.CassandraQueryTest
[INFO] Tests run: 3, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.086 s - in org.apache.flink.connector.cassandra.source.reader.CassandraQueryTest
[INFO] Tests run: 18, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 105.905 s - in org.apache.flink.streaming.connectors.cassandra.CassandraConnectorITCase
[INFO] 
[INFO] Results:
[INFO] 
[INFO] Tests run: 43, Failures: 0, Errors: 0, Skipped: 0
[INFO] 
[INFO] 
[INFO] --- japicmp-maven-plugin:0.16.0_m325:cmp (default) @ flink-connector-cassandra_2.12 ---
[INFO] Written file '/home/echauchot/projects/flink-connector-cassandra/flink-connector-cassandra/target/${rootDir}/tools/japicmp-output/flink-connector-cassandra_2.12/japicmp/japicmp.diff'.
[INFO] Written file '/home/echauchot/projects/flink-connector-cassandra/flink-connector-cassandra/target/${rootDir}/tools/japicmp-output/flink-connector-cassandra_2.12/japicmp/japicmp.xml'.
[INFO] Written file '/home/echauchot/projects/flink-connector-cassandra/flink-connector-cassandra/target/${rootDir}/tools/japicmp-output/flink-connector-cassandra_2.12/japicmp/japicmp.html'.
[INFO] 
[INFO] --- maven-install-plugin:2.5.2:install (default-install) @ flink-connector-cassandra_2.12 ---
[INFO] Installing /home/echauchot/projects/flink-connector-cassandra/flink-connector-cassandra/target/flink-connector-cassandra_2.12-4.0-SNAPSHOT.jar to /home/echauchot/.m2/repository/org/apache/flink/flink-connector-cassandra_2.12/4.0-SNAPSHOT/flink-connector-cassandra_2.12-4.0-SNAPSHOT.jar
[INFO] Installing /home/echauchot/projects/flink-connector-cassandra/flink-connector-cassandra/target/dependency-reduced-pom.xml to /home/echauchot/.m2/repository/org/apache/flink/flink-connector-cassandra_2.12/4.0-SNAPSHOT/flink-connector-cassandra_2.12-4.0-SNAPSHOT.pom
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  03:46 min
[INFO] Finished at: 2023-10-27T18:40:44+02:00
[INFO] ------------------------------------------------------------------------

```

echauchot replied:
> the idea which seems working at least for my setup with opensearch connector is using `<include>`, `<exclude>` for surefire plugin and different profiles like
> 
> ```
> <properties>
>     	<exclude.tests>nothing-to-exclude</exclude.tests>
> </properties>
> ...
> <plugin>
> 		<groupId>org.apache.maven.plugins</groupId>
> 		<artifactId>maven-surefire-plugin</artifactId>
> 		<configuration>
> 			...
> 		</configuration>
> 		<executions>
> 			<execution>
>                            <!--execute all the unit tests-->
> 			   <id>default-test</id>			
>                             <configuration>
>                             
> 	                            <includes>
> 		                            <include>${test.unit.pattern}</include>
> 	                            </includes>
> 	                            <excludes>
> 		                            <exclude>${exclude.tests}</exclude>
> 	                            </excludes>
>                                    ...
>                             </configuration>
>                       ...		
> 	              </execution>
> 	              <execution>
> 		      <!--execute all the integration tests-->
> 		      <id>integration-tests</id>
> 		      ...
> 		      <configuration>
> 			      <includes>
> 				      <include>**/*.*</include>
> 			      </includes>
> 			      <excludes>
> 				      <exclude>${test.unit.pattern}</exclude>
> 				      <exclude>${exclude.tests}</exclude>
> 			      </excludes>
> 			      ...
> 		      </configuration>
> 	      </execution>
>               </executions>
> </plugin>
> ...
> 	<profiles>
> 		<profile>
> 			<id>noArchUnit</id>
> 			<properties>
> 				<exclude.tests>**/org/apache/flink/architecture/**/*.java</exclude.tests>
> 			</properties>
> 		</profile>
> 	</profiles>
> ...
> ```

This solution is the first I proposed but it requires having a profile for exclusions in the pom of all the connectors so it is a breaking change and the solution was abandoned toward the -Dtest that does not require a change to the poms.
See my comment above -Dtest works as expected.

zentol replied:
> See my comment above -Dtest works as expected.

It doesn't. To be more explicit than @snuyanzin, this will fail in flink-sql-connector-* modules since they contain no tests. Naturally it works in the Cassandra repo because it doesn't have a sql-connector module, nor any module without any tests.

Surefire exclusions are the way to go, but we can't set these on the command-line since it'd override the unit/it test separation setup in the parent pom.
I'd suggest to add another property to the connector parent pom (e.g., `additionalExcludes`), which it adds to the surefire exclusions.
This should then "just work".

echauchot replied:
> It doesn't. To be more explicit than @snuyanzin, this will fail in flink-sql-connector-* modules since they contain no tests. Naturally it works in the Cassandra repo because it doesn't have a sql-connector module, nor any module without any tests.

Ah, the failure was about a module with no tests at all, it was written "without archunit tests". Indeed, in that case it does not work. Thanks for the precision Chesnay.

> Surefire exclusions are the way to go, but we can't set these on the command-line since it'd override the unit/it test separation setup in the parent pom.
I'd suggest to add another property to the connector parent pom (e.g., additionalExcludes), which it adds to the surefire exclusions.
This should then "just work".

Yes, at first I submitted [this PR](https://github.com/apache/flink-connector-shared-utils/pull/22/files) that added surefire exclusions to the parent pom by adding a custom profile but it had this drawback indeed. Of course I can add a property to add to already defined exclusions. The drawback I see that I wanted to avoid is that it modifies the parent pom leading to the need to release flink-connector-parent and to bump its version in the connectors. But I guess this trade-off is acceptable considering the sql modules problem above.

echauchot replied:
I have updated the PR as discussed above:
- use new **additionalExcludes** property to exclude archunit tests. This property is added in [this PR](https://github.com/apache/flink-connector-shared-utils/pull/26) that needs to be merged and released (v 1.0.1) before merging the current PR. See [here](https://github.com/apache/flink-connector-shared-utils/pull/26#issue-1992394911) how it was tested.
- move to **org.apache.flink**:flink-connector-parent instead of **io.github.zentol.flink**:flink-connector-parent in test project.
- changed the test project structure to include 2 submodules. Test-project is now a parent and there is one module with tests (the former test project)  and one without tests. This is to illustrate the case discussed above of the sql module containing no tests.

@zentol PTAL

zentol commented:
> > We could actually roll this into the run_dependency_convergence option because it has fundamentally the same cause.
> 
> But it seems I misunderstood what you meant. I actually also prefer to keep them separated. So, no problem, I'll make them separated

oh you understood me correctly, I just lead you in the wrong direction back then. So that's my fault 🙏

echauchot commented:
> @zentol I fixed the PR: mainly separated dep convergence checks and archunit tests.
> 
> I have also simulated an archunit violation with the behavior of the jira ticket: a rule was fixed in a snapshot leading to a new violation (ConnectorRules#CONNECTOR_CLASSES_ONLY_DEPEND_ON_PUBLIC_API that I fixed and that you reviewed). See [this commit](https://github.com/apache/flink-connector-shared-utils/pull/23/commits/eb28d4ba4f5434e356fd09120c9420f726004266). I think we can leave such a violation in the test project to serve for future tests.
> 

@zentol do you agree to keep this simulated violation for testing purpose ?

> After playing a lot with this simulation locally, I find that the comments in `archunit.properties` could be improved because not precise enough and even a bit misleading. I propose [this comments change](https://github.com/apache/flink-connector-shared-utils/pull/23/commits/a16852321a829fc0c42271570563655c41dc8d14). If you agree with the change, I'll do a PR to flink-architecture-tests module.

@zentol do you agree with my proposed modifications to the doc in the conf so that I can do the related flink-architecture-tests PR ?

echauchot commented:
> > @zentol I fixed the PR: mainly separated dep convergence checks and archunit tests.
> > I have also simulated an archunit violation with the behavior of the jira ticket: a rule was fixed in a snapshot leading to a new violation (ConnectorRules#CONNECTOR_CLASSES_ONLY_DEPEND_ON_PUBLIC_API that I fixed and that you reviewed). See [this commit](https://github.com/apache/flink-connector-shared-utils/pull/23/commits/eb28d4ba4f5434e356fd09120c9420f726004266). I think we can leave such a violation in the test project to serve for future tests.
> 
> @zentol do you agree to keep this simulated violation for testing purpose ?
> 
> > After playing a lot with this simulation locally, I find that the comments in `archunit.properties` could be improved because not precise enough and even a bit misleading. I propose [this comments change](https://github.com/apache/flink-connector-shared-utils/pull/23/commits/a16852321a829fc0c42271570563655c41dc8d14). If you agree with the change, I'll do a PR to flink-architecture-tests module.
> 
> @zentol do you agree with my proposed modifications to the doc in the conf so that I can do the related flink-architecture-tests PR ?

any comments on there 2 questions ?

zentol commented on module-with-tests/src/test/java/org/apache/flink/connector/testing/architecture/TestCodeArchitectureTest.java:
```diff
@@ -0,0 +1,46 @@
+
+/** Architecture tests for test code. */
+@AnalyzeClasses(
+        packages = {
+            "org.apache.flink.batch.connectors.cassandra",
+            "org.apache.flink.streaming.connectors.cassandra",
+            "org.apache.flink.connector.cassandra",
+            "org.apache.flink.connectors.cassandra"
+        },
```
These should cause the test to be a no-op.

echauchot replied:
oops, bad copy/paste. Thanks for pointing out !

echauchot replied:
Done

echauchot replied:
done

zentol commented on module-with-tests/src/main/java/org/apache/flink/connector/testing/SomeClass.java:
```diff
@@ -0,0 +1,31 @@
+public class SomeClass {
+    public static void main(String[] args) {
+        // this call is an intended (for testing) violation of the
+        // ConnectorRules#CONNECTOR_CLASSES_ONLY_DEPEND_ON_PUBLIC_API archunit rule that was fixed
```
I'd prefer a violation that is fully under the control of this repo.

zentol replied:
Actually, I'm a bit lost.  Does this fail the build or not? What prevents it from failing if the architecture tests are enabled? We always build against the same Flink version after all.

echauchot replied:
It was to simulate a rule recently changed in flink-snapshot (the initial problem that triggered this PR). When the archunit tests are enabled, if the violation is added to the store (via freeze) it should not fail the build and it should fail the build otherwise. But anyway as the general case is to disable archunit for both snapshots and non-main flink versions, I think it is better to remove the simulated violation indeed.

echauchot replied:
Done

zentol commented on pom.xml:
```diff
@@ -48,7 +52,41 @@ under the License.
+		<!-- ArchUit test dependencies -->
+		<dependency>
+			<groupId>org.apache.flink</groupId>
+			<artifactId>flink-architecture-tests-test</artifactId>
```
I'd rather have us define our own rules than relying on the Flink ones. We then wouldn't need these 2 dependencies, nor 10 violations stores, nor had to update them when Flink versions are being bumped.
For correctness purposes it shouldn't make a difference.

echauchot replied:
I agree, updates are a problem but I thought that test-project was a template for connectors projects. As, all the connectors rely on flink rules, shouldn't test project too ? 

If test-project is not intended to be a template for new connectors, but just a test project for the CI env, then I can definitely just create a single local rule for production classes and one for test classes. would you prefer that ?

zentol replied:
> I thought that test-project was a template for connectors projects

No, the Maven module in the ci_utils branch only exists for testing the workflow.

echauchot replied:
Done:
- added a simple archunit rule for both test and production code
- removed the flink archunit dependencies
- improved the dependencies declaration between parent and children modules

zentol commented on module-with-no-tests/.gitignore:
```diff
@@ -0,0 +1,38 @@
+!**/src/test/**/target/
+
+### IntelliJ IDEA ###
+.idea/modules.xml
```
if anything this belongs into a root .gitignore.

echauchot replied:
it is intellij auto-generated .gitignore.To be honest I did not even open it. I'll improve it. Thanks

echauchot replied:
Done

zentol commented on module-with-no-tests/src/main/java/org/apache/flink/connector/testing/SomeClass2.java:
```diff
@@ -19,4 +19,7 @@
 
 /** A dummy class; only exists so that the compile/packaging plugins have something to do. */
-public class SomeClass {}
+public class SomeClass2 {
```
Why is this class being modified?

echauchot replied:
not modified: the original class is still named SomeClass and is located in module-with-tests module (the former test-project). This SomeClass2 is another one for the module-with-no-tests module.

I can also have both classes named SomeClass and introduce different packages for module-with-tests and module-with-no-tests if you prefer

echauchot replied:
no more need for this SomeClass**2** name as the new archunit rule is on the name of the class, so the class in module-with-tests is now called SomeConnectorClass.

echauchot commented:
> > do you agree with my proposed modifications to the doc in the conf so that I can do the related flink-architecture-tests PR ?
> 
> Looks reasonable to me 👍
> 

Ok thanks, test-project template for new connectors is updated here and I update flink-architecture-tests flink-connector-base in [this PR](https://github.com/apache/flink/pull/23723) but I did not update all the externalized and non-externalized connectors.

> > do you agree to keep this simulated violation for testing purpose ?
> 
> I'm not too convinced it's worth the trouble. Right now it adds a lot of complexity, and afaict it may not work correctly.
> 

> It's fundamentally a tricky thing to test. What you kind of want is a test that always fails, so that you can check that the architecture exclusion is working correctly. 

exactly

> But you don't want the build to fail if the exclusion isn't present, because that's also part of the expected test matrix. 

if the exclusion is not present, the build should fail.

> I don't see a good way to test that that wouldn't introduce yet another parameter (including implicit ones like the Flink version), which then may invalidate the whole test if it's not working properly.

I don't want to introduce yet another parameter.

echauchot commented:
@zentol I addressed all your comments. PTAL.
please note: for the CI to pass, it needs a release of new connector-parent with the exclusion property.

echauchot commented:
Jira ticket moved to a subtask FLINK-34137: branch squashed and rebased and PR renamed.

echauchot commented:
@zentol:
Now that flink-connector-parent is released with skip archunit feature, I could update this CI PR to test this new feature:
- Used the new connector-parent
- Rebased on ci_utils
- upgraded all flinks versions in CI: 1.16.x -> 1.17.x, 1.17.x -> 1.18.x
PTAL

zentol commented on module-with-tests/.gitignore:
```diff
@@ -0,0 +1,34 @@
+.classpath
+.idea
+!.idea/vcs.xml
+.metadata
```
let's move this to the root directory

echauchot replied:
moved .metadata from modules **module-with-tests** and **module-with-no-tests** to **test-project-parent/.gitignore**

echauchot replied:
Is that the only thing ? Can I merge the PR ?

echauchot commented:
Fixed the ticket number in the commits, merging when the tests pass
